### PR TITLE
Improvement: refactoring and some performance optimizations for point cloud picking

### PR DESCRIPTION
**Pull request by Savokr**
# Description

Improvements for GPU point cloud picking:

- Resolution of picking frame matches resolution of Reveal.
- Refactored `PointCloudOctreePicker` class
- Added tests for `WebGLRendererStateHelper` and some fixes for the class
- Fixed docs example about reacting to clicks in point clouds

# Checklist:

- [x] I am proud of this feature.
- [x] I have performed a self-review of my own code.
- [x] I have manually tested this for different scenarios (different models, formats, devices, browsers).
- [x] I have commented my code in hard-to-understand areas.
- [x] I have made corresponding changes to the public documentation.
- [x] I have added unit and visuals tests to prove that my feature works to the best of my ability.
- [x] I have refactored the code for readability to the best of my ability.
- [x] I have checked that my changes do not introduce regressions in the public documentation.
- [x] I have outlined any known defects / lacking capabilities in the contents of this PR.
- [x] I have listed any remaining work that should follow this PR in the description or jira/miro/etc.
- [x] I have added TSDoc to any public facing changes.
- [x] I have added "developer documentation" in any package README.md that is related / applicable to this PR.
- [x] I have noted down and am currently tracking any technical debt introduced in this PR.
- [x] I have thoroughly thought about the architecture of this implementation.
- [x] I have asked peers to test this feature.

### CODE REVIEW

File: viewer/packages/pointclouds/src/PointCloudPickingHandler.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  private readonly _picker: PointCloudOctreePicker;
+
+  constructor(renderer: THREE.WebGLRenderer) {
+    this._picker = new PointCloudOctreePicker(renderer);
+  }
+
+  intersectPointClouds(
+    nodes: PointCloudNode[],
+    input: IntersectInput,
+    threshold: number = 0.05 // 5 cm
```
Comment (larsmoa) [inline viewer/packages/pointclouds/src/PointCloudPickingHandler.ts:47]: Does this have any effect?

File: viewer/packages/pointclouds/src/PointCloudPickingHandler.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    const { normalizedCoords, camera } = input;
+    this._normalized.set(normalizedCoords.x, normalizedCoords.y);
+    this._raycaster.setFromCamera(normalizedCoords, camera);
+    this._raycaster.params.Points = { threshold };
+
+    const intersections: PickPoint[] = [];
+
+    nodes.forEach(node => {
+      const intersection = this._picker.pick(camera, this._raycaster.ray, [node.potreeNode.octree], {
+        pickWindowSize: 20
```
Comment (larsmoa) [inline viewer/packages/pointclouds/src/PointCloudPickingHandler.ts:58]: What does 20 mean?

File: viewer/packages/pointclouds/src/potree-three-loader/Potree.ts [hunk 3]
```diff
@@ -111,8 +111,8 @@ export class Potree implements IPotree {
     ray: Ray,
     params: Partial<PickParams> = {}
   ): PickPoint | null {
-    Potree.picker = Potree.picker || new PointCloudOctreePicker();
-    return Potree.picker.pick(renderer, camera, ray, pointClouds, params);
+    Potree.picker = Potree.picker || new PointCloudOctreePicker(renderer);
```
Comment (larsmoa) [inline viewer/packages/pointclouds/src/potree-three-loader/Potree.ts:114]: Hmmmmmmmm! This looks scary, you commit to using the same renderer for all picking. What happens if you instantiate two viewers with different renderers and do point cloud picking?

Comment (larsmoa) [reply viewer/packages/pointclouds/src/potree-three-loader/Potree.ts:114]: Is `Potree.picker` a static global? If yes, is it really necessary?

**Review by larsmoa — CHANGES_REQUESTED** [review]
👍 for updated WebGLStateHelper and test.

`pointIntersectionThreshold` doesn't seem to be marked as deprecated in API Reference, which it should. Update tsdoc.

File: viewer/packages/pointclouds/src/PointCloudPickingHandler.ts (hunk 1 above)
Comment (Savokr) [reply viewer/packages/pointclouds/src/PointCloudPickingHandler.ts:47]: No, removed it as an argument

File: viewer/packages/pointclouds/src/PointCloudPickingHandler.ts (hunk 2 above)
Comment (Savokr) [reply viewer/packages/pointclouds/src/PointCloudPickingHandler.ts:58]: It's default PickingWindowSize that is a region on the screen around the cursor that will be checked for points there.

**Review by larsmoa — APPROVED** [review]
Approved, but see reply. I'm still a bit confused about `Potree.picker` and would kill it off is possible (and not to hard to do).

In general, we need to have a mindset that we improve the Potree code base step by step, which will probably mean doing extra work whenever we touch that code base...

File: viewer/packages/pointclouds/src/potree-three-loader/Potree.ts (hunk 3 above)
Comment (Savokr) [reply viewer/packages/pointclouds/src/potree-three-loader/Potree.ts]: Removed it

**Pull request merged**